pkg/backends/calico/routes_test.go

```diff
@@ -0,0 +1,138 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func addEndpointSubset(ep *v1.Endpoints, nodename string) {
+	ep.Subsets = append(ep.Subsets, v1.EndpointSubset{
+		Addresses: []v1.EndpointAddress{
+			v1.EndpointAddress{
+				NodeName: &nodename}}})
+}
+
+func buildSimpleService() (svc *v1.Service, ep *v1.Endpoints) {
+	meta := metav1.ObjectMeta{Namespace: "foo", Name: "bar"}
+	svc = &v1.Service{
+		ObjectMeta: meta,
+		Spec: v1.ServiceSpec{
+			Type:                  v1.ServiceTypeClusterIP,
+			ClusterIP:             "127.0.0.1",
+			ExternalTrafficPolicy: v1.ServiceExternalTrafficPolicyTypeLocal,
+		}}
+	ep = &v1.Endpoints{
+		ObjectMeta: meta,
+	}
+	return
+}
+
+var _ = Describe("RouteGenerator", func() {
+	var rg *routeGenerator
+	BeforeEach(func() {
+		rg = &routeGenerator{
+			nodeName:    "foobar",
+			svcIndexer:  cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil),
+			epIndexer:   cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil),
+			svcRouteMap: make(map[string]string),
+			client: &client{
+				cache: make(map[string]string),
+			},
+		}
+	})
+	Describe("getServiceForEndpoints", func() {
+		It("should get corresponding service for endpoints", func() {
+			// getServiceForEndpoints
+			svc, ep := buildSimpleService()
+			rg.svcIndexer.Add(svc)
+			fetchedSvc, key := rg.getServiceForEndpoints(ep)
+			Expect(fetchedSvc.ObjectMeta).To(Equal(svc.ObjectMeta))
+			Expect(key).To(Equal("foo/bar"))
+		})
+	})
+	Describe("getEndpointsForService", func() {
+		It("should get corresponding endpoints for service", func() {
+			// getEndpointsForService
+			svc, ep := buildSimpleService()
+			rg.epIndexer.Add(ep)
+			fetchedEp, key := rg.getEndpointsForService(svc)
+			Expect(fetchedEp.ObjectMeta).To(Equal(ep.ObjectMeta))
+			Expect(key).To(Equal("foo/bar"))
+		})
+	})
+
+	Describe("(un)setRouteForSvc", func() {
+		Context("svc = svc, ep = nil", func() {
+			It("should set an unset routes for a service", func() {
+				svc, ep := buildSimpleService()
+				addEndpointSubset(ep, rg.nodeName)
+
+				rg.epIndexer.Add(ep)
+				rg.setRouteForSvc(svc, nil)
+				Expect(rg.svcRouteMap["foo/bar"]).To(Equal("127.0.0.1/32"))
+				rg.unsetRouteForSvc(ep)
+				Expect(rg.svcRouteMap["foo/bar"]).To(BeEmpty())
+			})
+		})
+		Context("svc = nil, ep = ep", func() {
+			It("should set an unset routes for a service", func() {
+				svc, ep := buildSimpleService()
+				addEndpointSubset(ep, rg.nodeName)
+
+				rg.svcIndexer.Add(svc)
+				rg.setRouteForSvc(nil, ep)
+				Expect(rg.svcRouteMap["foo/bar"]).To(Equal("127.0.0.1/32"))
+				rg.unsetRouteForSvc(ep)
+				Expect(rg.svcRouteMap["foo/bar"]).To(BeEmpty())
+			})
+		})
+	})
+
+	Describe("resourceInformerHandlers", func() {
+		svc, ep := buildSimpleService()
+		BeforeEach(func() {
+			svc2 := *svc
+			ep2 := *ep
+			ep2.ObjectMeta.Name = svc2.ObjectMeta.Name
+			addEndpointSubset(&ep2, "barfoo")
+			rg.epIndexer.Add(ep2)
+
+			addEndpointSubset(ep, rg.nodeName)
```

Comment:
Also note here that you'll be adding a new endpoint subset to the same EP object for every test defined below - you probably want to declare a new EP for each test.

Reply:
Looks to me like that is now the case.